### DISCUSSION TITLE
求助(help)!connection refused!

### DISCUSSION THREAD

**Discussion by lamkei** [v2fly/v2ray-core#2489] [Q&A]
配置为vmess+ws+nginx+cdn，用了1年多了，5月1日突然无法使用，使用相同端口的伪装网站能够正常访问的。v2ray客户端的数据包能够到达nginx，但是nginx转发到v2ray的时候，error.log debug log记录connect() failed (111: Connection refused) while connecting to upstream。v2ray的log上没有任何记录产生。换了v2ray版本，或者更换为xray，或者换成非docker的v2ray都是一样的结果。甚至换了服务器和域名，结果还是没变。这是GFW污染了数据包还是怎样？求各位高手帮忙解决一下。(Dear guys! I used the v2ray for more than one years in the configuration of vmess + ws + nginx + cdn. But it crashed suddenly on 1 May. The pretended website can be visited normaly under the same nginx server, using the same port. So the data package from the v2ray client can reach the nginx, but the debug error log recorded the error that connect() failed (111: Connection refused) while connecting to upstream, while proxy passing the request to the v2ray server. And the log of v2ray did not record any thing, neither the access.log nor the error.log, under the debug level. I have tried to use another version of v2ray, use xray to replace, and use v2ray in none docker environmen, but got the same result. What's more, I tried to use another server and domain, but the result didn't change neither! Does the GFW pollute the data package or what ? Could you please help me?)

具体环境为(The actual environment is shown below): 
1. v2ray服务端: docker - v2fly/v2fly-core:v5.2.1，映射5000端口 (the v2ray server: using docker image: v2fly/v2fly-core:v5.2.1, and mapping the port 5000)

2. nginx服务器: docker nginx:latest，映射443端口 (the nginx server: using docker image nginx:latest, and mapping the port 443)

3. 创建docker bridge网络，分配v2ray: 192.168.1.2; nginx: 192.168.1.3 ( create a bridge docker network, and assigning 192.168.1.2 to v2ray, 192.168.1.3 to nginx)

4. cdn用cloudflare，ws路径跳过所有动作 ( using the cloudflare cdn, and set skip action while URI of incoming requests match the path set in v2ray server). 
5. v2ray 服务端配置(the configuration of v2ray server):

 {

        "log": {
                "loglevel": "debug",
                "access": "/var/log/v2ray/access.log",
                "error": "/var/log/v2ray/error.log"
         },
        "policy": {
                "levels": {
                        "0": {
                                "handshake": 4,
                                "connIdle": 3600,
                                "uplinkOnly": 2,
                                "downlinkOnly": 5,
                                "statusUserUplink": false,
                                "statusUserDownlink": false,
                                "bufferSize": 102400
                        }
                }
        },
        "inbounds": [{
                "port": 5000,
                "listen": "192.168.1.2",
                "protocol": "vmess",
                "settings": {
                    "clients": [
                        {
                            "id": "............................................",
                            "level": 0,
                            "alterId": 0
                        }
                    ]
                },
                "streamSettings": {
                    "network": "ws",
                    "wsSettings": {
                        "path": "/hello/world/man"
                    }
                }
        }],
        "outbounds": [
            {
                "protocol": "freedom",
                "settings": {}
            }
        ]
}

7. nginx配置(the configuration of nginx)

server {

	listen 443 ssl;
	server_name xxxxx.com;

	access_log /myLogs/access.log accessLog;
	error_log /myLogs/error.log debug;

	ssl_certificate /cert/fullchain.crt;
	ssl_certificate_key /cert/private.key;
	ssl_protocols TLSv1.2 TLSv1.3;
	ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384;

	location /hello/world/man {
		if ($http_upgrade != "websocket") {
			return 404;
		}
		proxy_redirect off;
		proxy_pass http://192.168.1.2:5000;
		proxy_http_version 1.1;
		proxy_set_header X-Forwarded-Host $host;
		proxy_set_header X-Forwarded-Server $host;
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header X-Forwarded-Proto $scheme;
		proxy_set_header Host $http_host;
		proxy_set_header Upgrade $http_upgrade;
		proxy_set_header Connection "upgrade";
		proxy_pass_request_headers on;
	}

	location / {
		alias /www/html/;
		index index.html;
	}
}
8. 客户端配置( the configuration of client named Qv2ray using the 4.45.0 v2ray core):

>基于4.45.0的Qv2ray的config 

{

    "outbounds": [
        {
            "_QV2RAY_USE_GLOBAL_FORWARD_PROXY_": false,
            "mux": {
                "concurrency": 1,
                "enabled": null
            },
            "protocol": "vmess",
            "sendThrough": "0.0.0.0",
            "settings": {
                "vnext": [
                    {
                        "address": " xxxxx.com",
                        "port": 443,
                        "users": [
                            {
                                "id": ".........................................................................."
                            }
                        ]
                    }
                ]
            },
            "streamSettings": {
                "network": "ws",
                "security": "tls",
                "tlsSettings": {
                    "alpn": [
                        "http/1.1"
                    ],
                    "serverName": " xxxxx.com"
                },
                "wsSettings": {
                    "headers": {
                        "host": " xxxxx.com"
                    },
                    "path": "/hello/world/man"
                },
                "xtlsSettings": {
                    "alpn": [
                        "http/1.1"
                    ],
                    "serverName": "xxxxx.com"
                }
            }
        }
    ]
}

9. nginx error log:

> [error] 20#20: *495 connect() failed (111: Connection refused) while connecting to upstream, client: xxx.xx.xx.xx, server: xxxxx.com, request: "GET /hello/world/man HTTP/1.1", upstream: "http://192.168.1.2:5000/hello/world/man", host: "xxxxx.com"

10. nginx access log:

> xxx.xx.xx.xx - - [02/May/2023:09:56:06 +0000] "GET /hello/world/man HTTP/1.1" 502 157 "-" "Go-http-client/1.1" "xxx.xx.xxx.xx" 192.168.1.2:5000 0.000 0.000

**Comment by mydogshitgold** [comment]
> [error] 20#20: *495 connect() failed (111: Connection refused) while connecting to upstream

Don't panic, and congrats, judging by the fact that you still got this NGINX log, you weren't blocked by some sketchy firewall... at least for now. And the log means that NGINX is unable to talk to the v2ray-core, which may be dead.
So let's first check if the v2ray daemon is up. what is the output of:
```
ss dst 192.168.1.2 dport 5000
```
and
```
pgrep -l v2ray
```


p.s. Could you reformat the body, it is unreadable.

**Reply by lamkei** [reply]
the ss dst 192.168.1.2 dport 5000 does not show any Processes, but pgrep can catch the v2ray process ID. 
But  how I finally solve the problem is that for you asked me to reformat the body, and as the github discussion formats itself and regard all blanks as one, I have to modify piece by piece to adapt for it. Then I found that, I put the inbounds in the policy while I missing a '}' in the v2ray server config, and the IDE does not report any problems as it is a correct json. !!!-_-. After correcting it, it is OK now. Thanks again!

**Reply by mydogshitgold** [reply]
No thanks necessary. It's good to know the problem is solved.
